### PR TITLE
optimize:optimize the auto-configuration processing of starter

### LINKED ISSUE #3440: The auto-configuration of seata starter should be optimized
The auto-configuration of seata starter should be optimized

### PR DESCRIPTION
**Pull request by xingfudeshi**
### Ⅰ. Describe what this PR did
The current implementation of seata auto-confuguration does not seem good.The PR does a optimization of it.

### Ⅱ. Does this pull request fix one issue?

fixes #3440 

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
I have tested it with spring-boot-mybatis

### Ⅳ. Describe how to verify it
Please test it with seata-samples(spring-boot-mybatis).

### CODE REVIEW

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * The type Seata properties configuration
+ *
+ * @author xingfudeshi@gmail.com
+ */
+@ConditionalOnProperty(prefix = SEATA_PREFIX, name = "enabled", havingValue = "true", matchIfMissing = true)
+@ComponentScan(basePackages = "io.seata.spring.boot.autoconfigure.properties")
+@AutoConfigureBefore({SeataAutoConfiguration.class, SeataDataSourceConfiguration.class})
+@Configuration
```
Comment (slievrly) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java:93]: if necessary?

Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java]: Yes.@Configuration can be optional if it was auto-configuration class.

Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java]: Fixed.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import static io.seata.spring.annotation.datasource.AutoDataSourceProxyRegistrar.BEAN_NAME_SEATA_DATA_SOURCE_BEAN_POST_PROCESSOR;
+import static io.seata.spring.boot.autoconfigure.StarterConstants.SEATA_PREFIX;
+
+/**
+ * The type Seata data source configuration.
+ *
+ * @author xingfudeshi@gmail.com
+ */
+@ConditionalOnProperty(prefix = SEATA_PREFIX, name = {"enableAutoDataSourceProxy", "enable-auto-data-source-proxy"}, havingValue = "true", matchIfMissing = true)
+public class SeataDataSourceConfiguration {
```
Comment (wangliang181230) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceConfiguration.java:35]: 我觉得应该以`AutoConfiguration`结尾吧。

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * The type Seata properties configuration
+ *
+ * @author xingfudeshi@gmail.com
+ */
+@ConditionalOnProperty(prefix = SEATA_PREFIX, name = "enabled", havingValue = "true", matchIfMissing = true)
+@ComponentScan(basePackages = "io.seata.spring.boot.autoconfigure.properties")
+@AutoConfigureBefore({SeataAutoConfiguration.class, SeataDataSourceConfiguration.class})
+public class SeataPropertiesConfiguration {
```
Comment (wangliang181230) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java:92]: 我觉得应该以`AutoConfiguration`结尾吧。

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceConfiguration.java (hunk 2 above)
Comment (a364176773) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceConfiguration.java:35]: +1

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java (hunk 3 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesConfiguration.java]: Fixed.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceConfiguration.java (hunk 2 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceConfiguration.java]: Fixed.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceAutoConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import static io.seata.spring.annotation.datasource.AutoDataSourceProxyRegistrar.BEAN_NAME_SEATA_AUTO_DATA_SOURCE_PROXY_CREATOR;
+import static io.seata.spring.annotation.datasource.AutoDataSourceProxyRegistrar.BEAN_NAME_SEATA_DATA_SOURCE_BEAN_POST_PROCESSOR;
+import static io.seata.spring.boot.autoconfigure.StarterConstants.SEATA_PREFIX;
+
+/**
+ * The type Seata data source auto configuration.
+ *
+ * @author xingfudeshi@gmail.com
+ */
+@ConditionalOnProperty(prefix = SEATA_PREFIX, name = {"enableAutoDataSourceProxy", "enable-auto-data-source-proxy"}, havingValue = "true", matchIfMissing = true)
```
Comment (wangliang181230) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceAutoConfiguration.java:34]: Does this configuration need to be disabled when the `seata.enabled=false`?

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesAutoConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import static io.seata.spring.boot.autoconfigure.StarterConstants.UNDO_PREFIX;
+
+/**
+ * The type Seata properties auto configuration
+ *
+ * @author xingfudeshi@gmail.com
+ */
+@ConditionalOnProperty(prefix = SEATA_PREFIX, name = "enabled", havingValue = "true", matchIfMissing = true)
+@ComponentScan(basePackages = "io.seata.spring.boot.autoconfigure.properties")
+@AutoConfigureBefore({SeataAutoConfiguration.class, SeataPropertiesAutoConfiguration.class})
```
Comment (wangliang181230) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataPropertiesAutoConfiguration.java:91]: `SeataAutoConfiguration` and `SeataDataSourceAutoConfiguration`.
Not the current class itself.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceAutoConfiguration.java (hunk 4 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceAutoConfiguration.java]: Yes.it's fixed now.

**Review by a364176773 — APPROVED** [review]
LGTM

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceAutoConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+import static io.seata.spring.annotation.datasource.AutoDataSourceProxyRegistrar.BEAN_NAME_SEATA_AUTO_DATA_SOURCE_PROXY_CREATOR;
+import static io.seata.spring.annotation.datasource.AutoDataSourceProxyRegistrar.BEAN_NAME_SEATA_DATA_SOURCE_BEAN_POST_PROCESSOR;
+
+/**
+ * The type Seata data source auto configuration.
+ *
+ * @author xingfudeshi@gmail.com
+ */
+@ConditionalOnExpression("${seata.enable:true} && ${seata.enableAutoDataSourceProxy:true} && ${seata.enable-auto-data-source-proxy:true}")
```
Comment (wangliang181230) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceAutoConfiguration.java:33]: `@ConditionalOnBean(DataSource.class)`？

Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataDataSourceAutoConfiguration.java:33]: done.

**Review by wangliang181230 — APPROVED** [review]
LGTM

**Pull request merged**